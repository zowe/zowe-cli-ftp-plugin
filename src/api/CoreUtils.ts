/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ICommandProfileTypeConfiguration, IImperativeError, Logger } from "@zowe/imperative";
import * as stream from "stream";
import { TransferMode } from "zos-node-accessor";

export class CoreUtils {

    /**
     * Sleep for a number of milliseconds
     * @param ms - how many miliseconds to sleep or (at least)
     * @returns promise that resolves when the sleep is complete (or after, not guaranteed)
     */
    public static async sleep(ms: number) {
        this.log.debug("Sleeping for %s milliseconds...", ms);
        return new Promise((resolve) => setTimeout(resolve, ms));
    }

    /**
     * Make sure newlines in a string have carriage returns
     * This is required to prevent errors when transferring
     * in ascii mode
     * @param  original - the original string which you would like to ensure has \r before \n
     * @returns a string where every \n has a \r before it
     */
    public static addCarriageReturns(original: string): string {
        this.log.debug("Adding carriage returns to string of length %s", original.length);
        return original.replace(/\r?\n/g, "\r\n");
    }

    /**
     * Read the complete contents of stdin
     * @param stream - A custom stream to read stdin data from
     * @returns contents piped in to stdin
     */
    public static async readStdin(stream?: stream.Readable): Promise<Buffer> {
        return new Promise<Buffer>((resolve, reject) => {
            let stdinContent: Buffer = Buffer.from([]);
            stream = stream || process.stdin;
            stream.resume();

            stream.on("data", (chunk: Buffer) => {
                this.log.trace(`Read data from stdin: ${chunk.toString()}`);
                this.log.debug(`Read ${chunk.length} bytes of data from stdin`);
                stdinContent = Buffer.concat([stdinContent, chunk]);
            });

            stream.once("end", () => {
                this.log.info("Finished reading stdin");
                resolve(stdinContent);
            });

            stream.on("error", (error: Error) => {
                const stdinReadError: IImperativeError = {
                    msg: "Error encountered while reading from stdin",
                    causeErrors: error,
                    additionalDetails: error?.message
                };
                reject(stdinReadError);
            });
        });
    }

    public static async getProfileSchema(): Promise<ICommandProfileTypeConfiguration[]> {
        const ftpProfile = await require("../imperative").profiles as ICommandProfileTypeConfiguration[];
        return ftpProfile;
    }

    protected static get log(): Logger {
        return Logger.getAppLogger();
    }

    public static getBinaryTransferModeOrDefault(isBinary: boolean, rdw = false): TransferMode {
        return (isBinary ?
            (rdw ? TransferMode.BINARY_RDW : TransferMode.BINARY) :
            (rdw ? TransferMode.ASCII_RDW : TransferMode.ASCII)
        );
    }
}
