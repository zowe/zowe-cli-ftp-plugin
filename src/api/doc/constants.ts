/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { TransferMode } from "zos-node-accessor";

/**
 * Bytes per track.
 */
export const TRACK = 56664;

/**
 * The data is transferred in text mode, in which encoding conversion like ASCII/EBCDIC will happen.
 */
export const TRANSFER_TYPE_ASCII = TransferMode.ASCII;

/**
 * The data is transferred in binary mode, in which no encoding conversion will happen.
 */
export const TRANSFER_TYPE_BINARY = TransferMode.BINARY;

/**
 * The data is transferred in text mode like TRANSFER_TYPE_ASCII, and 4-byte RDW is inserted at beginning of each record.
 */
export const TRANSFER_TYPE_ASCII_RDW = TransferMode.ASCII_RDW;
/**
  * The data is transferred in binary mode like TRANSFER_TYPE_BINARY, and 4-byte RDW is inserted at beginning of each record.
  */
export const TRANSFER_TYPE_BINARY_RDW = TransferMode.BINARY_RDW;