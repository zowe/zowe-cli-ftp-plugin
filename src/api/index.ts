/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

export * as zosNodeAccessor from "zos-node-accessor";
export { TransferMode } from "zos-node-accessor";

export * from "./doc";

export * from "./CoreUtils";
export * from "./FTPConfig";
export * from "./JobUtils";
export * from "./DataSetUtils";
export * from "./UssUtils";
