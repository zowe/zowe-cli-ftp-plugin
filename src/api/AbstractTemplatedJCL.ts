/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { IO, Logger, TextUtils } from "@zowe/imperative";
import { CoreUtils } from "./CoreUtils";

export abstract class AbstractTemplatedJCL {

    /**
     * Default job template for submitting your JCL
     */
    public abstract readonly DEFAULT_TEMPLATE: string;

    /**
     *
     * @param jobCardFile - file containing a job card which will be prepended to the job
     * @param  substitutionValues -
     * @param  templateFile - override the default JCL template with a file
     */
    public getJcl(jobCardFile: string, substitutionValues: {[key: string]: string}, templateFile?: string) {
        let jcl = IO.readFileSync(jobCardFile).toString();
        if (!/\r?\nZ/.test(jcl)) {
            // add a newline to the end of the jcl if there is none
            jcl += "\r\n";
        }
        if (templateFile != null) {
            // we
            jcl += IO.readFileSync(templateFile).toString();
        } else {
            jcl += this.DEFAULT_TEMPLATE;
        }
        return TextUtils.renderWithMustache(CoreUtils.addCarriageReturns(jcl), substitutionValues);
    }

    protected get log(): Logger {
        return Logger.getAppLogger();
    }
}
