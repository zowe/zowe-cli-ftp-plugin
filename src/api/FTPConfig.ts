/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ZosAccessor } from "zos-node-accessor";
import { ConnectionPropsForSessCfg, ICommandArguments, ICommandOptionDefinition, IHandlerParameters, ISession } from "@zowe/imperative";
import { IConnectionOption } from "./doc";

const ftpConnectionOptionGroup: string = "FTP Connection options";

const tlsConnectionOptionGroup: string = "TLS / Secure Connection options";

export class FTPConfig {

    public static OPTION_HOST: ICommandOptionDefinition = {
        type: "string",
        name: "host", aliases: ["H"],
        description: "The hostname or IP address of the z/OS server to connect to.",
        group: ftpConnectionOptionGroup
    };

    public static OPTION_PORT: ICommandOptionDefinition = {
        type: "number",
        name: "port", aliases: ["P"],
        description: "The port of the z/OS FTP server.",
        defaultValue: 21,
        group: ftpConnectionOptionGroup
    };

    public static OPTION_USER: ICommandOptionDefinition = {
        type: "string",
        name: "user", aliases: ["u"],
        description: "Username for authentication on z/OS",
        group: ftpConnectionOptionGroup
    };

    public static OPTION_PASSWORD: ICommandOptionDefinition = {
        type: "string",
        name: "password", aliases: ["p", "pass", "pw"],
        description: "Password to authenticate to FTP.",
        group: ftpConnectionOptionGroup
    };

    public static OPTION_SECURE_FTP: ICommandOptionDefinition = {
        name: "secure-ftp",
        type: "boolean",
        defaultValue: true,
        description: "Set to true for both control and data connection encryption," +
            " 'control' for control connection encryption only, or 'implicit' for implicitly" +
            " encrypted control connection (this mode is deprecated in modern times, but usually uses port 990). " +
            "Note: Unfortunately, this plugin's functionality only works with FTP and FTPS, not 'SFTP' which is FTP over SSH.",
        group: ftpConnectionOptionGroup
    };

    public static OPTION_REJECT_UNAUTHORIZED: ICommandOptionDefinition = {
        name: "reject-unauthorized",
        aliases: ["ru"],
        description: "Reject self-signed certificates. Only specify this if " +
            "you are connecting to a secure FTP instance.",
        type: "boolean",
        defaultValue: null,
        group: tlsConnectionOptionGroup
    };

    public static OPTION_SERVER_NAME: ICommandOptionDefinition = {
        name: "server-name",
        aliases: ["sn"],
        description: "Server name for the SNI (Server Name Indication) TLS extension. " +
            "Only specify if you are connecting securely",
        group: tlsConnectionOptionGroup,
        type: "string",
        defaultValue: null
    };

    public static OPTION_CONNECTION_TIMEOUT: ICommandOptionDefinition = {
        name: "connection-timeout", aliases: ["ct"],
        description: "How long (in milliseconds) to wait for the control connection to be established.",
        defaultValue: 10000,
        type: "number",
        group: ftpConnectionOptionGroup
    };


    public static OPTION_ENCODING: ICommandOptionDefinition = {
        name: "encoding", aliases: ["ec"],
        description: "The encoding for download and upload of z/OS data set.",
        defaultValue: null,
        type: "string",
        group: ftpConnectionOptionGroup
    };

    /**
     * Shared options used by all FTP commands
     */
    public static FTP_CONNECTION_OPTIONS: ICommandOptionDefinition[] = [
        FTPConfig.OPTION_HOST,
        FTPConfig.OPTION_PORT,
        FTPConfig.OPTION_USER,
        FTPConfig.OPTION_PASSWORD,
        FTPConfig.OPTION_SECURE_FTP,
        FTPConfig.OPTION_REJECT_UNAUTHORIZED,
        FTPConfig.OPTION_SERVER_NAME,
        FTPConfig.OPTION_CONNECTION_TIMEOUT
    ];

    /**
     * Convert a profile into a config object used to connect to
     * zos-node-accessor
     * @param arguments - the arguments passed by the user
     * @param doPrompting - Whether to prompt for missing arguments (defaults to true)
     * @param {IHandlerParameters} handlerParams - The command parameters object for daemon mode prompting
     * @returns  the connection to zos-node-accessor's APIs
     */
    public static async connectFromArguments(args: ICommandArguments, doPrompting = true, handlerParams?: IHandlerParameters): Promise<ZosAccessor> {
        const sessCfg: ISession = {
            type: "basic",
            hostname: args.host,
            port: args.port,
            user: args.user,
            password: args.password
        };
        const sessCfgWithCreds = await ConnectionPropsForSessCfg.addPropsOrPrompt(sessCfg, args, { doPrompting, parms: handlerParams });
        const ftpConfig = FTPConfig.createConfigFromArguments({
            ...args,
            host: sessCfgWithCreds.hostname,
            port: sessCfgWithCreds.port,
            user: sessCfgWithCreds.user,
            password: sessCfgWithCreds.password
        });
        const zosAccessor = new ZosAccessor();
        return zosAccessor.connect(ftpConfig);
    }

    /**
     * Convert a profile into a config object used to connect to
     * zos-node-accessor
     * @param arguments - the profile created by the user
     * @returns the config object for zos-node-accessor's ftp connection
     */
    public static createConfigFromArguments(args: Partial<ICommandArguments>): IConnectionOption {
        // build the options argument for zos-node-accessor
        const result: IConnectionOption = {
            host: args.host,
            user: args.user,
            password: args.password,
            port: args.port,
            connTimeout: args.connectionTimeout
        };
        if (args.secureFtp != null) {
            if (typeof args.secureFtp === "string" && args.secureFtp.trim().toLowerCase() === "true") {
                result.secure = true;
            } else {
                result.secure = args.secureFtp;
            }
        }
        if (this.profileHasSecureOptions(args)) {
            result.secureOptions = {
                rejectUnauthorized: args.rejectUnauthorized,
                serverName: args.serverName,
            };
        }
        return result;
    }

    /**
     * Determine whether the user specified any TLS connection options
     * @param args - the profile loaded from the user
     * @returns true if the user has specified any secure connection options
     */
    private static profileHasSecureOptions(args: Partial<ICommandArguments>): boolean {
        return (args.rejectUnauthorized != null) || (args.serverName !== null);
    }
}
