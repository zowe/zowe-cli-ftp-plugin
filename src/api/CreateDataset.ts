/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { CreateDataSetTypeEnum, CreateDefaults, ICreateDataSetOptions, IZosFilesResponse, ZosFilesMessages } from "@zowe/cli";
import { ImperativeError, ImperativeExpect, TextUtils } from "@zowe/imperative";
import { JobUtils } from "./JobUtils";
import { CoreUtils } from "./CoreUtils";
import { AbstractTemplatedJCL } from "./AbstractTemplatedJCL";

export class CreateDataset extends AbstractTemplatedJCL {

    /**
     * Create a data set
     * @param connection                 - zos-node-accessor connection
     * @param  cmdType               - The type of data set we are going to create
     * @param dataSetName                          - the name of the data set to create
     * @param  [options={}] - additional options for the creation of the data set
     */
    public static async create(connection: any, cmdType: CreateDataSetTypeEnum,
        dataSetName: string,
        jobCardFile: string,
        options ?: Partial<ICreateDataSetOptions>):
        Promise<IZosFilesResponse> {
        let validCmdType = true;

        // Removes undefined properties
        let tempOptions = options != null ? JSON.parse(JSON.stringify(options)) : {};
        const secondarySpecified = tempOptions.secondary != null;

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(cmdType, ZosFilesMessages.missingDatasetType.message);

        // Required
        ImperativeExpect.toNotBeNullOrUndefined(dataSetName, ZosFilesMessages.missingDatasetName.message);

        switch (cmdType) {
            case CreateDataSetTypeEnum.DATA_SET_PARTITIONED:
                tempOptions = {...CreateDefaults.DATA_SET.PARTITIONED, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_SEQUENTIAL:
                tempOptions = {...CreateDefaults.DATA_SET.SEQUENTIAL, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_BINARY:
                tempOptions = {...CreateDefaults.DATA_SET.BINARY, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_C:
                tempOptions = {...CreateDefaults.DATA_SET.C, ...tempOptions};
                break;
            case CreateDataSetTypeEnum.DATA_SET_CLASSIC:
                tempOptions = {...CreateDefaults.DATA_SET.CLASSIC, ...tempOptions};
                break;
            default:
                validCmdType = false;
                break;
        }

        if (!validCmdType) {
            throw new ImperativeError({msg: ZosFilesMessages.unsupportedDatasetType.message});
        } else {
            // Handle the size option
            if (tempOptions.size != null) {
                const tAlcunit = tempOptions.size.toString().match(/[a-zA-Z]+/g);
                if (tAlcunit != null) {
                    tempOptions.alcunit = tAlcunit.join("").toUpperCase();
                }

                const tPrimary = tempOptions.size.toString().match(/[0-9]+/g);
                if (tPrimary != null) {
                    tempOptions.primary = +(tPrimary.join(""));

                    if (!secondarySpecified) {
                        const TEN_PERCENT = 0.10;
                        tempOptions.secondary = Math.round(tempOptions.primary * TEN_PERCENT);
                    }
                }

                delete tempOptions.size;
            }

            let response = "";
            // Handle the print attributes option
            if (tempOptions.printAttributes != null) {
                if (tempOptions.printAttributes) {
                    delete tempOptions.printAttributes;
                    response = TextUtils.prettyJson(tempOptions);
                } else {
                    delete tempOptions.printAttributes;
                }
            }

            response = await new CreateDataset().createViaFTP(connection, dataSetName, tempOptions, jobCardFile);
            if (response.indexOf("RC=0000") >= 0) {
                return {
                    success: true,
                    commandResponse: response + "\n" + ZosFilesMessages.dataSetCreatedSuccessfully.message,
                }
            } else {
                return {
                    success: false,
                    commandResponse: response + "\nFailed to create data set",
                }
            }
        }
    }


    public DEFAULT_TEMPLATE = "//CREATE EXEC PGM=IEFBR14\n" +
        "//{{createDD}}";

    private async createViaFTP(connection: any, dataSetName: string,
        options: ICreateDataSetOptions,
        jobCardFile: string,
        overrideTemplateFile ?: string) {
        this.log.debug("Building jcl to create data set %s", dataSetName);

        let createDD: string = "NEWDATA DD ";
        const createDDValues = ["DSN=" + dataSetName.toUpperCase(), "DISP=(NEW,CATLG)"];
        const maxDDLineLength = 65;
        let currentLineLength = createDD.length;
        const continueLine = "//   ";

        // todo: make sure all options are used here
        if (options.alcunit) {
            let spaceParams = "SPACE=(";
            spaceParams += options.alcunit.toUpperCase() + ",(";
            spaceParams += options.primary;
            if (options.secondary != null) {
                spaceParams += "," + options.secondary;
            }
            spaceParams += "))";

            createDDValues.push(spaceParams);
            if (options.volser != null) {
                createDDValues.push("VOLSER=" + options.volser.toUpperCase());
            }
        }
        if (options.lrecl != null) {
            createDDValues.push("LRECL=" + options.lrecl);
        }
        if (options.recfm != null) {
            createDDValues.push("RECFM=" + options.recfm.toUpperCase());
        }
        if (options.blksize != null) {
            createDDValues.push("BLKSIZE=" + options.blksize);
        }
        if (options.unit != null) {
            createDDValues.push("UNIT=" + options.unit.toUpperCase());
        }

        if (options.dsorg != null) {
            createDDValues.push("DSORG=" + options.dsorg.toUpperCase());
        }

        /**
         * Build the DD definition for the job by looping through the values we've built
         */
        for (let i = 0; i < createDDValues.length; i++) {
            const value = createDDValues[i];
            if (value.length + currentLineLength > maxDDLineLength) {
                createDD += "\n" + continueLine;
                currentLineLength = continueLine.length;
            }

            createDD += value;
            if (i < createDDValues.length - 1) {
                // if this isn't the last value, add a comma to continue the line
                createDD += ",";
                currentLineLength += 1;
            }
            currentLineLength += value.length;
        }

        const jcl = this.getJcl(jobCardFile, {createDD}, overrideTemplateFile);
        this.log.debug("Creating data set %s via JCL:\n %s", dataSetName, jcl);
        const jobid = await connection.submitJCL(jcl);
        const jobWait = 500;
        await CoreUtils.sleep(jobWait); // have to wait for the job to show up in the list
        const jobDetails = await JobUtils.findJobByID(connection, jobid);
        return "Job details: " + JSON.stringify(jobDetails, null, 2);
    }

}
