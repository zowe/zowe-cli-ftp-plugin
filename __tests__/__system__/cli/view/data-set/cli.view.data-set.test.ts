/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 *
 */

import { ITestEnvironment, TestEnvironment, runCliScript } from "@zowe/cli-test-utils";
import { ITestPropertiesSchema } from "../../../../__src__/doc/ITestPropertiesSchema";
import { FTPConfig } from "../../../../../src/api/FTPConfig";
import { generateRandomAlphaNumericString, generateRandomBytes } from "../../../../__src__/TestUtils";

let connection: any;
let testDataSet: string;
let testEnvironment: ITestEnvironment<ITestPropertiesSchema>;

describe("view data-set command", () => {
    // Create the unique test environment
    beforeAll(async () => {
        testEnvironment = await TestEnvironment.setUp({
            tempProfileTypes: ["zftp"],
            testName: "zos_view_dataset",
            installPlugin: true
        });
        expect(testEnvironment).toBeDefined();
        connection = await FTPConfig.connectFromArguments(testEnvironment.systemTestProperties.zftp);
        testDataSet = testEnvironment.systemTestProperties.datasets.writablePDS;
    });

    afterAll(async () => {
        connection.close();
        await TestEnvironment.cleanUp(testEnvironment);
    });

    it("should be able to view the job data set from the test properties file", async () => {
        const iefbr14DataSet = testEnvironment.systemTestProperties.jobs.iefbr14Member.toUpperCase();
        const iefbr14Content = (await connection.getDataset(iefbr14DataSet)).toString();
        const result = runCliScript(__dirname + "/__scripts__/command_view_data_set.sh", testEnvironment, [iefbr14DataSet]);
        expect(result.stderr.toString()).toEqual("");
        expect(result.status).toEqual(0);
        expect(result.stdout.toString().trim()).toContain("IEFBR14");
        expect(result.stdout.toString().trim()).toContain(iefbr14Content.trim());
    });

    it("should be able to upload to a data set in binary mode, then view the content with the CLI and " +
        "verify that the content is correct", async () => {
        const randomContentLength = 60;
        const randomContent = await generateRandomBytes(randomContentLength);
        const memberSuffixLength = 6;
        const destination = testDataSet + "(R" + generateRandomAlphaNumericString(memberSuffixLength) + ")";
        await connection.uploadDataset(randomContent, "'" + destination + "'", "binary");
        const result = runCliScript(__dirname + "/__scripts__/command_view_data_set_binary.sh", testEnvironment,
            [destination]);
        expect(result.stderr.toString()).toEqual("");
        expect(result.status).toEqual(0);
        const uploadedContent = (await connection.getDataset("'" + destination + "'", "binary"));
        // binary upload to a fixed record  data set will fill a data set with zeroes for the remainder of the record
        // so we can trim the zeroes off and still be accurate
        const uploadedContentString = uploadedContent.toString("hex").replace(/0+$/g, "");
        expect(uploadedContentString).toEqual(randomContent.toString("hex").replace(/0+$/g, ""));
        await connection.deleteDataset(destination);
    });

    it("should give a syntax error if the data set is omitted", async () => {
        const result = runCliScript(__dirname + "/__scripts__/command_view_data_set.sh", testEnvironment, []);
        const stderr = result.stderr.toString();
        expect(stderr).toContain("Positional");
        expect(stderr).toContain("data set");
        expect(stderr).toContain("Syntax");
        expect(result.status).toEqual(1);
    });
});
