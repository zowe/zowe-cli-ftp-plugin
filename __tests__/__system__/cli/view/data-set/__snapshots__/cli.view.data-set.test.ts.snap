// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view data-set command should display view data set help 1`] = `
"================z/OS VIEW LIST DATA-SET HELP===============

 COMMAND NAME
 ------------

   data-set | ds

 DESCRIPTION
 -----------

   View the contents of a z/OS data set

 USAGE
 -----

   zowe zos-ftp view data-set <dataSet> [options]

 POSITIONAL ARGUMENTS
 --------------------

   dataSet		 (string)

      The data set (PDS member or physical sequential data set) which you would like
      to view the contents of.

 OPTIONS
 -------

   --binary  | -b (boolean)

      View content in binary form without converting to ASCII text

   --secure-ftp  (string)

      Set to true for both control and data connection encryption, 'control' for
      control connection encryption only, or 'implicit' for implicitly encrypted
      control connection (this mode is deprecated in modern times, but usually uses
      port 990). Note: Unfortunately, this plugin's functionality only works with FTP
      and FTPS, not 'SFTP' which is FTP over SSH.

   --connection-timeout  | --ct (number)

      How long (in milliseconds) to wait for the control connection to be established.

      Default value: 10000

 REQUIRED OPTIONS
 ----------------

   --host  | -H (string)

      The hostname or IP address of the z/OS server to connect to.

   --port  | -P (number)

      The port of the z/OS FTP server.

      Default value: 21

   --user  | -u (string)

      Username for authentication on z/OS

   --password  | --pass | --pw (string)

      Password to authenticate to FTP.

 TLS / SECURE CONNECTION OPTIONS
 -------------------------------

   --reject-unauthorized  | --ru (boolean)

      Reject self-signed certificates. Only specify this if you are connecting to a
      secure FTP instance.

   --server-name  | --sn (string)

      Server name for the SNI (Server Name Indication) TLS extension. Only specify if
      you are connecting securely

 PROFILE OPTIONS
 ---------------

   --zftp-profile  | --zftp-p (string)

      The name of a (zftp) profile to load for this command execution.

 GLOBAL OPTIONS
 --------------

   --response-format-json  | --rfj (boolean)

      Produce JSON formatted data from a command

   --help  | -h (boolean)

      Display help text


 EXAMPLES
 --------

   - View the content of the data set \\"ibmuser.cntl(iefbr14)\\":

      $ zowe zos-ftp view data-set \\"ibmuser.cntl(iefbr14)\\"

   - View the content of the data set \\"ibmuser.loadlib(main)\\"
   and pipe it into the hex viewer program xxd:

      $ zowe zos-ftp view data-set \\"ibmuser.loadlib(main)\\" -b | xxd

================z/OS FTP VIEW DATA-SET HELP WITH RFJ===========
{
  \\"success\\": true,
  \\"exitCode\\": 0,
  \\"message\\": \\"The help was constructed for command: data-set.\\",
  \\"stdout\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   data-set | ds\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   View the contents of a z/OS data set\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   zowe zos-ftp view data-set <dataSet> [options]\\\\n\\\\n POSITIONAL ARGUMENTS\\\\n --------------------\\\\n\\\\n   dataSet\\\\t\\\\t (string)\\\\n\\\\n      The data set (PDS member or physical sequential data set) which you would like\\\\n      to view the contents of.\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --binary  | -b (boolean)\\\\n\\\\n      View content in binary form without converting to ASCII text\\\\n\\\\n   --secure-ftp  (string)\\\\n\\\\n      Set to true for both control and data connection encryption, 'control' for\\\\n      control connection encryption only, or 'implicit' for implicitly encrypted\\\\n      control connection (this mode is deprecated in modern times, but usually uses\\\\n      port 990). Note: Unfortunately, this plugin's functionality only works with FTP\\\\n      and FTPS, not 'SFTP' which is FTP over SSH.\\\\n\\\\n   --connection-timeout  | --ct (number)\\\\n\\\\n      How long (in milliseconds) to wait for the control connection to be established.\\\\n\\\\n      Default value: 10000\\\\n\\\\n REQUIRED OPTIONS\\\\n ----------------\\\\n\\\\n   --host  | -H (string)\\\\n\\\\n      The hostname or IP address of the z/OS server to connect to.\\\\n\\\\n   --port  | -P (number)\\\\n\\\\n      The port of the z/OS FTP server.\\\\n\\\\n      Default value: 21\\\\n\\\\n   --user  | -u (string)\\\\n\\\\n      Username for authentication on z/OS\\\\n\\\\n   --password  | --pass | --pw (string)\\\\n\\\\n      Password to authenticate to FTP.\\\\n\\\\n TLS / SECURE CONNECTION OPTIONS\\\\n -------------------------------\\\\n\\\\n   --reject-unauthorized  | --ru (boolean)\\\\n\\\\n      Reject self-signed certificates. Only specify this if you are connecting to a\\\\n      secure FTP instance.\\\\n\\\\n   --server-name  | --sn (string)\\\\n\\\\n      Server name for the SNI (Server Name Indication) TLS extension. Only specify if\\\\n      you are connecting securely\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zftp-profile  | --zftp-p (string)\\\\n\\\\n      The name of a (zftp) profile to load for this command execution.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   - View the content of the data set \\\\\\"ibmuser.cntl(iefbr14)\\\\\\":\\\\n\\\\n      $ zowe zos-ftp view data-set \\\\\\"ibmuser.cntl(iefbr14)\\\\\\"\\\\n\\\\n   - View the content of the data set \\\\\\"ibmuser.loadlib(main)\\\\\\"\\\\n   and pipe it into the hex viewer program xxd:\\\\n\\\\n      $ zowe zos-ftp view data-set \\\\\\"ibmuser.loadlib(main)\\\\\\" -b | xxd\\\\n\\\\n\\",
  \\"stderr\\": \\"\\",
  \\"data\\": \\"\\\\n COMMAND NAME\\\\n ------------\\\\n\\\\n   data-set | ds\\\\n\\\\n DESCRIPTION\\\\n -----------\\\\n\\\\n   View the contents of a z/OS data set\\\\n\\\\n USAGE\\\\n -----\\\\n\\\\n   zowe zos-ftp view data-set <dataSet> [options]\\\\n\\\\n POSITIONAL ARGUMENTS\\\\n --------------------\\\\n\\\\n   dataSet\\\\t\\\\t (string)\\\\n\\\\n      The data set (PDS member or physical sequential data set) which you would like\\\\n      to view the contents of.\\\\n\\\\n OPTIONS\\\\n -------\\\\n\\\\n   --binary  | -b (boolean)\\\\n\\\\n      View content in binary form without converting to ASCII text\\\\n\\\\n   --secure-ftp  (string)\\\\n\\\\n      Set to true for both control and data connection encryption, 'control' for\\\\n      control connection encryption only, or 'implicit' for implicitly encrypted\\\\n      control connection (this mode is deprecated in modern times, but usually uses\\\\n      port 990). Note: Unfortunately, this plugin's functionality only works with FTP\\\\n      and FTPS, not 'SFTP' which is FTP over SSH.\\\\n\\\\n   --connection-timeout  | --ct (number)\\\\n\\\\n      How long (in milliseconds) to wait for the control connection to be established.\\\\n\\\\n      Default value: 10000\\\\n\\\\n REQUIRED OPTIONS\\\\n ----------------\\\\n\\\\n   --host  | -H (string)\\\\n\\\\n      The hostname or IP address of the z/OS server to connect to.\\\\n\\\\n   --port  | -P (number)\\\\n\\\\n      The port of the z/OS FTP server.\\\\n\\\\n      Default value: 21\\\\n\\\\n   --user  | -u (string)\\\\n\\\\n      Username for authentication on z/OS\\\\n\\\\n   --password  | --pass | --pw (string)\\\\n\\\\n      Password to authenticate to FTP.\\\\n\\\\n TLS / SECURE CONNECTION OPTIONS\\\\n -------------------------------\\\\n\\\\n   --reject-unauthorized  | --ru (boolean)\\\\n\\\\n      Reject self-signed certificates. Only specify this if you are connecting to a\\\\n      secure FTP instance.\\\\n\\\\n   --server-name  | --sn (string)\\\\n\\\\n      Server name for the SNI (Server Name Indication) TLS extension. Only specify if\\\\n      you are connecting securely\\\\n\\\\n PROFILE OPTIONS\\\\n ---------------\\\\n\\\\n   --zftp-profile  | --zftp-p (string)\\\\n\\\\n      The name of a (zftp) profile to load for this command execution.\\\\n\\\\n GLOBAL OPTIONS\\\\n --------------\\\\n\\\\n   --response-format-json  | --rfj (boolean)\\\\n\\\\n      Produce JSON formatted data from a command\\\\n\\\\n   --help  | -h (boolean)\\\\n\\\\n      Display help text\\\\n\\\\n\\\\n EXAMPLES\\\\n --------\\\\n\\\\n   - View the content of the data set \\\\\\"ibmuser.cntl(iefbr14)\\\\\\":\\\\n\\\\n      $ zowe zos-ftp view data-set \\\\\\"ibmuser.cntl(iefbr14)\\\\\\"\\\\n\\\\n   - View the content of the data set \\\\\\"ibmuser.loadlib(main)\\\\\\"\\\\n   and pipe it into the hex viewer program xxd:\\\\n\\\\n      $ zowe zos-ftp view data-set \\\\\\"ibmuser.loadlib(main)\\\\\\" -b | xxd\\\\n\\\\n\\"
}"
`;
